### ISSUE TITLE
Invalid suggestion for single_match

### ISSUE THREAD

**Issue by Keats** [rust-lang/rust-clippy#810]
It gives me the following:

``` rust
src/lib.rs:112:17: 115:18 warning: you seem to be trying to use match for destructuring a single pattern. Consider using `if let`, #[warn(single_match)] on by default
src/lib.rs:112                 match self.$field_name {
src/lib.rs:113                     Some(ref value) => { d.insert(stringify!($field_name).to_string(), value.to_json()); },
src/lib.rs:114                     None => {},
src/lib.rs:115                 }
src/lib.rs:118:9: 118:24 note: in this expansion of optional! (defined in src/lib.rs)
src/lib.rs:112:17: 115:18 help: try this
src/lib.rs:                    if let Some(ref value) = self.$field_name {
src/lib.rs:                        Some(ref value) => { d.insert(stringify!($field_name).to_string(), value.to_json()); },
src/lib.rs:                        None => {},
src/lib.rs:                    }
src/lib.rs:                )
src/lib.rs:            }
```

while the suggestion should be (afaik):

``` rust
                if let Some(ref value) = self.$field_name {
                    d.insert(stringify!($field_name).to_string(), value.to_json());
                }
```

**Comment by mcarton** [comment]
Weird, probably because of the macro, we get the span wrong.

**Comment by phansch** [comment]
Going to close this in favor of #1404 since it's most likely the same issue and has some more code examples.
